x-pack/plugin/identity-provider/qa/idp-rest-tests/src/test/java/org/elasticsearch/xpack/idp/IdentityProviderAuthenticationIT.java

```diff
@@ -0,0 +1,240 @@
+        final Map<String, Object> request = Map.ofEntries(
+            Map.entry("name", "Test SP"),
+            Map.entry("acs", SP_ACS),
+            Map.entry("privileges", Map.ofEntries(
+                Map.entry("resource", SP_ENTITY_ID),
+                Map.entry("roles", Map.of("superuser", "role:superuser", "viewer", "role:viewer"))
+            )),
+            Map.entry("attributes", Map.ofEntries(
+                Map.entry("principal", "https://idp.test.es.elasticsearch.org/attribute/principal"),
+                Map.entry("name", "https://idp.test.es.elasticsearch.org/attribute/name"),
+                Map.entry("email", "https://idp.test.es.elasticsearch.org/attribute/email"),
+                Map.entry("roles", "https://idp.test.es.elasticsearch.org/attribute/roles")
+            )));
+        final SamlServiceProviderIndex.DocumentVersion docVersion = createServiceProvider(SP_ENTITY_ID, request);
+        checkIndexDoc(docVersion);
+        ensureGreen(SamlServiceProviderIndex.INDEX_NAME);
+        final String samlResponse = generateSamlResponse(SP_ENTITY_ID, SP_ACS, null);
+        authenticateWithSamlResponse(samlResponse, null);
+    }
+
+    public void testRegistrationAndSpInitiatedSso() throws Exception {
+        final Map<String, Object> request = Map.ofEntries(
+            Map.entry("name", "Test SP"),
+            Map.entry("acs", SP_ACS),
+            Map.entry("privileges", Map.ofEntries(
+                Map.entry("resource", SP_ENTITY_ID),
+                Map.entry("roles", Map.of("superuser", "role:superuser", "viewer", "role:viewer"))
+            )),
+            Map.entry("attributes", Map.ofEntries(
+                Map.entry("principal", "https://idp.test.es.elasticsearch.org/attribute/principal"),
+                Map.entry("name", "https://idp.test.es.elasticsearch.org/attribute/name"),
+                Map.entry("email", "https://idp.test.es.elasticsearch.org/attribute/email"),
+                Map.entry("roles", "https://idp.test.es.elasticsearch.org/attribute/roles")
+            )));
+        final SamlServiceProviderIndex.DocumentVersion docVersion = createServiceProvider(SP_ENTITY_ID, request);
+        checkIndexDoc(docVersion);
+        ensureGreen(SamlServiceProviderIndex.INDEX_NAME);
+        Tuple<String, String> idAndRedirect = generateSamlAuthnRequest(REALM_NAME);
+        final String requestId = idAndRedirect.v1();
+        final String query = idAndRedirect.v2().split("\\?")[1];
+        Map<String, Object> authnState = validateAuthnRequest(SP_ENTITY_ID, query);
+        final String samlResponse = generateSamlResponse(SP_ENTITY_ID, SP_ACS, authnState);
+        assertThat(samlResponse, containsString("InResponseTo=\"" + requestId + "\""));
+        authenticateWithSamlResponse(samlResponse, requestId);
+    }
+
+    private void checkIndexDoc(SamlServiceProviderIndex.DocumentVersion docVersion) throws IOException {
```

Comment:
I think we can pull this into the base class from here and `ManageServiceProviderRestIT`